packages/connect-node/src/node-universal-client.spec.ts

```diff
@@ -40,118 +40,802 @@ describe("Node.js http2 API", function () {
           {}
         );
         setTimeout(() => {
-          stream.close(rstCode, () => {
-            session.close();
-            setTimeout(() => resolve(), 0);
+          stream.close(http2.constants.NGHTTP2_CANCEL, () => {
+            // We are seeing a race condition in Node v16.20.0, where closing
+            // the session right after closing a stream with an RST code
+            // _sometimes_ sends an INTERNAL_ERROR code.
+            // Simply delaying the session close until the next tick like
+            // we do here seems to work around the issue.
+            // We do _not_ guard against this case in the universal client,
+            // since we were not able to reproduce the issue there.
+            setTimeout(() => session.close(resolve), 0);
           });
         }, 0);
       });
     });
-  }
+
+    while (serverReceivedRstCode === undefined) {
+      // wait for the server to see the reset code
+      await new Promise((resolve) => setTimeout(resolve, 1));
+    }
+    expect(serverReceivedRstCode).toBe(http2.constants.NGHTTP2_CANCEL);
+  });
 });
 
-describe("universal node http2 client", function () {
-  describe("with a signal that is already aborted", function () {
-    it("should raise error with code canceled", async function () {
-      const signal = AbortSignal.abort();
-      const client = createNodeHttpClient({
-        httpVersion: "2",
-        baseUrl: "http://example.com",
-        keepSessionAlive: false,
-      });
-      try {
-        await client({
-          url: "http://example.com",
+describe("universal node http client", function () {
```

Comment:
These added tests spin up a Node.js HTTP/1.1 or HTTP/2.2 server that aborts incoming requests. We run a universal Node.js client (which is used by all Transports from `@bufbuild/connect-node`) against the servers. We test closing the request at several stages, for example before the first response byte is written, or while the request body is still read.